---
title: "Custom reference data in GenomicDistributions"
author: "MichaÅ‚ Stolarczyk"
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{1. Custom reference data in GenomicDistributions}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, echo=FALSE}
# These settings make the vignette prettier
knitr::opts_chunk$set(results="hold", collapse=FALSE, message=FALSE)
#refreshPackage("GenomicDistributions")
#devtools::build_vignettes("code/GenomicDistributions")
#devtools::test("code/GenomicDistributions")
```

# Introduction

`GenomicDistributions` package comes with a limited reference data set. There are only files that support `hg19` human reference genome. Other reference data can be generated with functions included in this package.

# Generate non-standard reference data set

In this vignette we'll go over the steps to generate reference data for a non-standard genome assembly: *C. Elegans* (Caenorhabditis_elegans.WBcel235.103).

All we need to generate the custom reference data set are two local or remote files. We will use remote files from Ensembl FTP server:

1. GTF - [http://ftp.ensembl.org/pub/release-103/gtf/caenorhabditis_elegans/Caenorhabditis_elegans.WBcel235.103.gtf.gz](http://ftp.ensembl.org/pub/release-103/gtf/caenorhabditis_elegans/Caenorhabditis_elegans.WBcel235.103.gtf.gz)
2. FASTA - [http://ftp.ensembl.org/pub/release-103/fasta/caenorhabditis_elegans/dna/Caenorhabditis_elegans.WBcel235.dna.toplevel.fa.gz](http://ftp.ensembl.org/pub/release-103/fasta/caenorhabditis_elegans/dna/Caenorhabditis_elegans.WBcel235.dna.toplevel.fa.gz)

```{r}
fastaSource = "http://ftp.ensembl.org/pub/release-103/fasta/caenorhabditis_elegans/dna/Caenorhabditis_elegans.WBcel235.dna.toplevel.fa.gz"
gtfSource = "http://ftp.ensembl.org/pub/release-103/gtf/caenorhabditis_elegans/Caenorhabditis_elegans.WBcel235.103.gtf.gz"
```

## Chromosome Sizes

```{r, echo=FALSE}
library(GenomicDistributions)
```

```{r message=TRUE}
CEelegansChromSizes = getChromSizesFromFasta(source=fastaSource)
```
Let's inspect the result:
```{r}
CEelegansChromSizes
```


## Transcription Start Sites (TSS)

```{r message=TRUE}
CEelegansTSSs = getTssFromGTF(source=gtfSource, convertEnsemblUCSC=TRUE)
```

Let's inspect the result:
```{r}
head(CEelegansTSSs)
```

## Gene Models

```{r message=TRUE}
features = c("gene", "exon", "three_prime_utr", "five_prime_utr")
CEelegansGeneModels = getGeneModelsFromGTF(source=gtfSource, features=features, convertEnsemblUCSC=TRUE)
```

Let's inspect the result:
```{r}
head(CEelegansGeneModels)
```